### PR TITLE
feat(web): redirect to login on expired cookie

### PR DESCRIPTION
**Pull request by ludviglundgren**
Redirect to `login` on expired cookie session.

Fixes #201

### CODE REVIEW

File: web/src/api/APIClient.ts [hunk 1]
```diff
@@ -22,7 +23,13 @@ export async function HttpClient<T>(
 
     return window.fetch(`${baseUrl()}${endpoint}`, config)
         .then(async response => {
-            if ([401, 403, 404].includes(response.status))
+            if (response.status === 401) {
+                AuthContext.set([{ username: "", isLoggedIn: false }] as any)
```
[the remark this answers is not in the document]
Comment (stacksmash76) [reply web/src/api/APIClient.ts:27]: I'd also clear the cookies first by doing:
```
cookies = new Cookies();
cookies.remove("user_session");
```
and afterwards resetting the AuthContext object.

Cookies can be imported from react-cookies: ```import { Cookies } from "react-cookie";```

File: web/src/App.tsx [hunk 2]
```diff
@@ -9,29 +9,44 @@ import Logout from "./screens/auth/logout";
 import { baseUrl } from "./utils";
 
 import { AuthContext, SettingsContext } from "./utils/Context";
+import {Fragment, useEffect} from "react";
+import {APIClient} from "./api/APIClient";
+import {useCookies} from "react-cookie";
 
 function Protected() {
+    const [authContext, setAuthContext] = AuthContext.use();
```
Comment (stacksmash76) [inline web/src/App.tsx:17]: I'd say that all the changes made in this file are unneeded. I've tested the solution proposed in my two comments from below and it works as expected -- on an invalid cookie, upon the first next network activity, the user gets logged out and the cookie gets cleared.

**Review by stacksmash76 — CHANGES_REQUESTED** [review]
Needs a few adjustments. Changes made to App.tsx seem not to be needed.

There's a rather small problem with our solution: the promise returned by HttpClient needs to be cancelled when it returns the control flow to react-query. When the promise gets returned, react-query will try to refresh the component which made this request, but it will be unable to since firstly the parent (i.e. App) component will get refreshed, which will unload the child HttpClient-caller component before react-query gets to return data to it, at which point -- kaboom:
```
Error: Unauthorized
    at APIClient.ts:32:1
react_devtools_backend.js:3973 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    at Base (http://localhost:3000/static/js/bundle.js:7844:79)
    at Protected
```
   
See https://juliangaramendy.dev/blog/use-promise-subscription and https://gist.github.com/pygy/6290f78b078e22418821b07d8d63f111. IIRC, I think react-error-boundary handles this without leakage, but not sure. Might be a bandaid fix to a bullet wound.

**Comment by ludviglundgren** [comment]
Made the suggested changes and that simplifies it. It still gives warning about `update in unmounted component`, but thats fine.

My first solution with the check to validate might have been a bit over-engineered for this case but the overall pattern of that isn't bad.

**Review by stacksmash76 — APPROVED** [review]
Looks good to me!

**Pull request merged**